webview-ui/src/components/settings/SettingsView.tsx

```diff
@@ -299,11 +309,26 @@ const SettingsView = forwardRef<SettingsViewRef, SettingsViewProps>(({ onDone, o
 
 	useImperativeHandle(ref, () => ({ checkUnsaveChanges }), [checkUnsaveChanges])
 
-	const onConfirmDialogResult = useCallback((confirm: boolean) => {
-		if (confirm) {
+	// kilocode_change start
+	const onConfirmDialogResult = useCallback(
+		(confirm: boolean) => {
+			if (confirm) {
+				setCachedState(extensionState)
+				setChangeDetected(false)
+				setDiscardInProgress(true)
+			}
+		},
+		[setCachedState, setChangeDetected, extensionState],
+	)
+
+	// Call the stored callback after discarding is complete
+	useEffect(() => {
+		if (discardInProgress === true) {
 			confirmDialogHandler.current?.()
+			setDiscardInProgress(false)
 		}
-	}, [])
+	}, [discardInProgress])
```

Comment:
This whole structure feels a bit weird. If you need to do something after other callbacks are completed, why not just a `setTimeout(() => confirmDialogHandler.current?.(), 0)`? in the place of `setDiscardInProgress(true)`?